subo/command/docs.go

```diff
@@ -0,0 +1,171 @@
+				return errors.Wrap(err, "failed to GetString")
+			} else if err := generateDocs(dir, outputDir, goBuildFlags); err != nil {
+				return errors.Wrap(err, "failed to getUpdatedDocs")
+			}
+
+			return nil
+		},
+	}
+	cmd.Flags().String("output", "", "output directory for generated documentation")
+	cmd.MarkFlagRequired("output")
+
+	return cmd
+}
+
+// DocsTestCmd returns the docs test command
+func DocsTestCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "test [dir] -- [go build flags]",
+		Short: "Test code and snippets inserts",
+		Long:  `Test code and snippets inserts without generating new documentation`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var goBuildFlags []string
+			dir := "."
+			if len(args) > 0 {
+				dir = args[0]
+				goBuildFlags = args[1:]
+			}
+
+			if cache, err := util.CacheDir(); err != nil {
+				return errors.Wrap(err, "failed to CacheDir")
+			} else if err := generateDocs(dir, filepath.Join(cache, "docs"), goBuildFlags); err != nil {
+				return errors.Wrap(err, "failed to getUpdatedDocs")
+			}
+
+			return nil
+		},
+	}
+
+	return cmd
+}
+
+func generateDocs(dir string, outputDir string, goBuildFlags []string) error {
+	filePaths, err := getVerifiedFiles(dir, goBuildFlags)
+	if err != nil {
+		return errors.Wrap(err, "failed to getVerifiedFiles")
+	}
+
+	for fileName, fileContent := range filePaths["md"] {
+		filePath := filepath.Join(outputDir, fileName)
+		fileDir := filepath.Dir(filePath)
+		if info, err := os.Stat(fileDir); os.IsNotExist(err) {
+			if err := os.MkdirAll(fileDir, os.ModePerm); err != nil {
+				return errors.Wrap(err, "failed to MkdirAll")
+			}
+		} else if err != nil {
+			return errors.Wrap(err, "failed to Stat")
+		} else if !info.IsDir() {
+			return errors.New(fmt.Sprintf("%s is not a directory", fileDir))
+		}
+
+		if file, err := os.Create(filePath); err != nil {
+			return errors.Wrap(err, "failed to Create")
+		} else if tmpl, err := template.New(fileName).Parse(fileContent); err != nil {
+			return errors.Wrap(err, "failed to Parse")
+		} else if err = tmpl.Execute(file, filePaths); err != nil {
+			return errors.Wrap(err, "failed to Execute")
+		}
+	}
+
+	return nil
+}
+
+func getVerifiedFiles(dir string, goBuildFlags []string) (map[string]map[string]string, error) {
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		return nil, errors.Wrap(err, fmt.Sprintf("dir %s does not exist", dir))
+	} else if err != nil {
+		return nil, errors.Wrap(err, "failed to Stat")
+	}
+
+	// go build
+	goBuildArgs := []string{"build"}
+	if goBuildFlags != nil {
+		// go build [-o output] [build flags]
+		goBuildArgs = append(goBuildArgs, goBuildFlags...)
+	}
+	// go build [-o output] [build flags] [packages]
+	goBuildArgs = append(goBuildArgs, []string{dir}...)
+
+	if output, err := exec.Command("go", goBuildArgs...).CombinedOutput(); err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("failed to build code due to:\n%s", output))
+	}
```

Comment:
Go already has a nice idiomatic way of testing code snippets as part of the `go doc` tool along with `go test`. Doing it this way, we can get Examples sections for free in `go docs` as well.

I haven't looked too far into it, but I'm pretty sure we can use parts of the `go doc` codebase to extract snippets of code and use that instead of rolling our own file system crawling stuff.

Let's talk about this more! I have a few ideas.

Reply:
See: https://go.dev/blog/examples

Reply:
@jagger27 I went through both `go doc` and `godoc` code base and only `godocs` actually extracts the testable examples and formats everything (no static doc api either) nicely like in that [shared example](https://go.dev/blog/examples).

The good news is that I converted `subo docs` to leverage the same code extraction as `godocs`. Unfortunately, we can't easily leverage godocs formatting since those apis are baked into godocs server logic (no static doc apis).

Our formatting code is actually pretty basic since we only want the testable examples and not everything that `godoc` formatting offers, so it's pretty light. So I'm not sure there's much benefit in adapting godocs formatting
since most of the formatting code is private and baked into godoc server logic.

Let me know if you like this formatting or not. I updated the main example up above.